### PR TITLE
[Webserver] Use operator_name instead of task_type in UI

### PR DESCRIPTION
**Pull request by utkarsharma2**
we should use `operator_name` when displaying operator name in UI.

Related: #28466 (not closing yet, see discussion below)

### CODE REVIEW

File: airflow/www/views.py [hunk 1]
```diff
@@ -3067,7 +3067,7 @@ class GraphForm(DateTimeWithNumRunsWithDagRunsForm):
         tasks = {
             t.task_id: {
                 "dag_id": t.dag_id,
-                "task_type": t.task_type,
+                "operator_name": t.operator_name,
```
Comment (uranusjr) [inline airflow/www/views.py]: This is passed to `graph.html` and dumped into JSON for JavaScript. I think you also need to change something in JavaScript to match this change.

Comment (utkarsharma2) [reply airflow/www/views.py]: @uranusjr I removed it, it turns out we don't need to change that part of the code. 
The only usage of `task_type` is at https://github.com/apache/airflow/blob/3bcaf3dd7e71993e1a89cf9367bf7c540cc88e75/airflow/www/static/js/graph.js#L211 Which I think should remain intact.

**Review by uranusjr — APPROVED** [review]

**Comment by utkarsharma2** [comment]
Found one more instance of `task_type` usage in UI, working on changes accordingly.

**Comment by utkarsharma2** [comment]
@uranusjr While going through the code I found there are two other tooltips where we show the operator name 
https://github.com/apache/airflow/blob/29d2a31dc04471fc92cbfb2943ca419d5d8a6ab0/airflow/www/static/js/task_instances.js#L111
and 
https://github.com/apache/airflow/blob/29d2a31dc04471fc92cbfb2943ca419d5d8a6ab0/airflow/www/static/js/task_instances.js#L170

Ideally, both of them should show values derived from `custom_operator_name` but only the base operator has the `operator_name` property and not the task instance object. In both of the instance as mentioned above, we are displaying values of `ti.operator` which are populated from `task_type` of operator.
 https://github.com/apache/airflow/blob/29d2a31dc04471fc92cbfb2943ca419d5d8a6ab0/airflow/models/taskinstance.py#L888

since `self.operator` is getting used at multiple places in code, I think we should create another property in the task instance operator similar to `operator_name` in the base operator and use that while displaying in UI. WDYT?

**Comment by uranusjr** [comment]
Hmm, yeah adding a new field is probably the simplest approach. Can you find other uses of `TaskInstance.operator` so maybe we can analyse if this field is simply for UI (if so, we can just change it to use `operator_name`), or do other logic depend on it? Adding a field is relatively expensive on TaskInstance so we should avoid that if possible.

Either way, I think we should hold this PR as-is and continue looking into `TaskInstance.operator` in a different PR.

File: airflow/www/views.py [hunk 2]
```diff
@@ -3084,6 +3084,7 @@ class GraphForm(DateTimeWithNumRunsWithDagRunsForm):
             t.task_id: {
                 "dag_id": t.dag_id,
                 "task_type": t.task_type,
```
Comment (uranusjr) [inline airflow/www/views.py:3086]: Do we still need this? If the only use case if to check SubDagOperator, we can use operator_name as well.

**Comment by utkarsharma2** [comment]
> Hmm, yeah adding a new field is probably the simplest approach. Can you find other uses of `TaskInstance.operator` so maybe we can analyze if this field is simply for UI (if so, we can just change it to use `operator_name`), or do other logic depends on it? Adding a field is relatively expensive on TaskInstance so we should avoid that if possible.
> 
> Either way, I think we should hold this PR as-is and continue looking into the `TaskInstance.operator` in a different PR.

@uranusjr I checked for the usage of `ti.operator` and found out there are a lot of places where it's been used, below are a few examples  
https://github.com/apache/airflow/blob/8891c6944671d8965a63ddbd59b5c23597e86f0a/airflow/models/dagrun.py#L1126
https://github.com/apache/airflow/blob/8891c6944671d8965a63ddbd59b5c23597e86f0a/airflow/models/dag.py#L1749
https://github.com/apache/airflow/blob/8891c6944671d8965a63ddbd59b5c23597e86f0a/airflow/jobs/scheduler_job_runner.py#L744

`TaskInstance.operator` certainly has more usage than just displaying in UI.

I think adding another field to `TaskInstance` will be a quick solution. WDYT?

File: airflow/www/views.py (hunk 2 above)
Comment (utkarsharma2) [reply airflow/www/views.py:3086]: Yes, I think we still need this since the below code compares the operator class name and it's best to keep this since the operato_name can change.
https://github.com/apache/airflow/blob/8891c6944671d8965a63ddbd59b5c23597e86f0a/airflow/www/static/js/graph.js#L211
and based on that displaying the subdag button -
https://github.com/apache/airflow/blob/8891c6944671d8965a63ddbd59b5c23597e86f0a/airflow/www/static/js/callModal.js#L185-L190

**Comment by uranusjr** [comment]
Yes that’s probably best. I think some of the use cases can technically switch to `operator_name` (e.g. logging), but at least the second one you mentioned should use the raw class name.

File: airflow/www/views.py (hunk 2 above)
Comment (bbovenzi) [reply airflow/www/views.py:3086]: Wouldn't it be better to pass `is_subdag` from the webserver instead? That's what we do in the REST API

Comment (utkarsharma2) [reply airflow/www/views.py:3086]: Yes, we can. I'll update the code. :)

Comment (utkarsharma2) [reply airflow/www/views.py:3086]: @bbovenzi Would it make sense to raise a separate PR for this change? because I think it's a separate task.

**Comment by utkarsharma2** [comment]
@uranusjr Added a new property `custom_operator_name` in TaskInstance class.

Also, would be nice if you can cross verify all the places that should be populating the value for `custom_operator_name` are covered.

**Comment by utkarsharma2** [comment]
@uranusjr since there is a new field added to `TaskInstance` class which requires a new column to be created in `TaskInstance` table. I think this would be a breaking change, right? Do we have something that does database migration  similar to [Django migrations](https://docs.djangoproject.com/en/4.2/topics/migrations/)?  if so, we can add a default value so that the database migrations can be smooth for existing TaskInstance.

**Comment by potiuk** [comment]
> @uranusjr since there is a new field added to `TaskInstance` class which requires a new column to be created in `TaskInstance` table. I think this would be a breaking change, right? Do we have something that does database migration similar to [Django migrations](https://docs.djangoproject.com/en/4.2/topics/migrations/)? if so, we can add a default value so that the database migrations can be smooth for existing TaskInstance.

We have alembic. You can see the docs about addin new migrations (look for alembic) and you can also take a look at past prs that added migrations to `airflow\migrations`. Also worth looking at the `airflow db` command and all their subcommands and https://airflow.apache.org/docs/apache-airflow/stable/migrations-ref.html

File: airflow/models/taskinstance.py [hunk 3]
```diff
@@ -394,6 +394,7 @@ class TaskInstance(Base, LoggingMixin):
     queue = Column(String(256))
     priority_weight = Column(Integer)
     operator = Column(String(1000))
+    custom_operator_name = Column(String(1000))
```
Comment (uranusjr) [inline airflow/models/taskinstance.py]: Let’s split changes related to ti to another PR as well, this involves migration and complicates the changes quite a lot.

Comment (utkarsharma2) [reply airflow/models/taskinstance.py]: @uranusjr Sure, I'll move it to other PR

Comment (utkarsharma2) [reply airflow/models/taskinstance.py]: Moved all the `taskinstance` related changes to PR - https://github.com/apache/airflow/pull/31852

**Review by uranusjr — APPROVED** [review]

File: airflow/www/views.py (hunk 2 above)
Comment (jedcunningham) [reply airflow/www/views.py:3086]: Yeah, let's do a separate PR.

**Review by jedcunningham — APPROVED** [review]
